README.md
# servant-tracing


ChangeLog.md
# Changelog for servant-tracing

## Unreleased changes


test/Spec.hs

import Servant.TracingSpec (tracingProps, tracingSpecs)
import Zipkin.ClientSpec (zipkinProps, zipkinSpec)
import Tracing.CoreSpec (coreSpec, coreProps)
import Test.Tasty

main :: IO ()
main = defaultMain $ testGroup "Tracing"  [
    tracingProps,
    tracingSpecs,
    zipkinProps,
    zipkinSpec,
    coreSpec
    ]


test/Instances.hs
{-# LANGUAGE NamedFieldPuns, OverloadedStrings, DuplicateRecordFields #-}

module Instances where

import Servant.Tracing
import Tracing.Core

import Data.Char (isAscii)
import Test.QuickCheck hiding (sample)
import Data.Time.Clock.POSIX
import qualified Data.Map as M
import qualified Data.Text as T

instance Arbitrary TracingInstructions where
    arbitrary = do
        tid <- arbitrary
        sid <- arbitrary
        parentSpanId <- arbitrary
        dbg <- arbitrary
        sample <- arbitrary
        pure TracingInstructions {
            traceId=tid,
            spanId=sid,
            parentSpanId,
            debug=dbg,
            sample
            }
        where

positiveArb :: (Integral a, Arbitrary a) => Gen a
positiveArb = suchThat arbitrary (>= 0)

instance Arbitrary TraceId where
    arbitrary = TraceId <$> positiveArb

instance Arbitrary SpanId where
    arbitrary = SpanId <$> positiveArb

instance Arbitrary SpanContext where
    arbitrary = SpanContext <$> arbitrary <*> arbitrary

instance Arbitrary OpName where
    arbitrary = OpName <$> arbitrary

instance Arbitrary SpanRelation where
    arbitrary = do
        rel <- elements [ChildOf, FollowsFrom]
        rel <$> arbitrary

instance Arbitrary T.Text where
    arbitrary =
        T.pack <$> listOf (suchThat arbitrary isAscii)

instance Arbitrary SpanRelationTag where
    arbitrary = elements [Child, Follows]

-- TODO add tags and baggage once they're supported
instance Arbitrary Span where
    arbitrary = do
        o <- arbitrary
        c <- arbitrary
        rels <- arbitrary
        dbg <- arbitrary
        svc <- arbitrary
        pure Span {
            operationName = o,
            context = c,
            timestamp = 1522024571,
            duration = 123,
            relations = rels,
            tags = M.empty,
            baggage = M.empty,
            debug = dbg,
            serviceName = svc
            }



test/Tracing/CoreSpec.hs
{-# LANGUAGE NamedFieldPuns, OverloadedStrings, UndecidableInstances, DuplicateRecordFields, FlexibleInstances #-}

module Tracing.CoreSpec (
    coreProps,
    coreSpec
    ) where

import Tracing.Core
import Instances()

import Control.Monad.Trans (liftIO, MonadIO)
import Control.Monad.Trans.Control (MonadBaseControl)
import Control.Monad.Reader
import Data.Maybe (fromMaybe, maybe, isJust)
import Data.Aeson
import Data.Aeson.Types (parseMaybe)
import Data.IORef (IORef, newIORef, readIORef)
import qualified Data.Text as T
import Test.HUnit
import Test.QuickCheck hiding (sample)
import Test.Tasty
import Test.Tasty.HUnit (testCase)
import Test.Tasty.QuickCheck (testProperty)
import Web.HttpApiData (parseUrlPiece)

coreProps :: TestTree
coreProps = testGroup "Tracing Core Properties" $ [
    ]

coreSpec :: TestTree
coreSpec = testGroup "Tracing Core Specification" $ [
    testCase "logged span id == fresh id for top level span" $ do
        ctx <- newContext
        flip runReaderT ctx . recordSpan Nothing "Top Level Test" $ pure (7 + 8)
        traces <- readIORef . spanBuffer $ tracer ctx
        let h = head traces
        length traces @=? 1
        relations h @=? []
        (sid $ context h) @=? SpanId 1234
    , testCase "logged span id == new id for child span" $ do
        ctx <- newContext
        flip runReaderT ctx . recordSpan (Just Child) "Child Test" $ pure (7 + 8)
        traces <- readIORef . spanBuffer $ tracer ctx
        let h = head traces
        length traces @=? 1
        (length $ relations h) @=? 1
        ((sid $ context h) /= SpanId 1234)  @? "Parent Id is still the child span's Id"
    , testCase "logged span id == new id for successor span" $ do
        ctx <- newContext
        flip runReaderT ctx . recordSpan (Just Follows) "Successor Test" $ pure (7 + 8)
        traces <- readIORef . spanBuffer $ tracer ctx
        let h = head traces
        length traces @=? 1
        ((sid $ context h) /= SpanId 1234)  @? "Parent Id is still the successor span's Id"
    , testCase "child span sets parent relation" $ do
        ctx <- newContext
        flip runReaderT ctx . recordSpan (Just Child) "Foo" $ pure (7 + 8)
        traces <- readIORef . spanBuffer $ tracer ctx
        let h = head traces
        length traces @=? 1
        ((\(ChildOf s) -> sid s ) . head $ relations h) @=?  SpanId 1234
    , testCase "successor span sets precursor relation" $ do
        ctx <- newContext
        flip runReaderT ctx . recordSpan (Just Follows) "Foo" $ pure (7 + 8)
        traces <- readIORef . spanBuffer $ tracer ctx
        let h = head traces
        ((\(FollowsFrom s) -> sid s ) . head $ relations h) @=?  SpanId 1234
    , testCase "sibling spans share same parent" $ do
        ctx <- newContext
        flip runReaderT ctx . recordSpan (Just Child) "Foo" $ pure (7 + 8)
        flip runReaderT ctx . recordSpan (Just Child) "Foo" $ pure (7 + 8)
        [x,y] <- readIORef . spanBuffer $ tracer ctx
        (parentId x == parentId y) @? "Sibling spans must share a parent"
    , testCase "sibling successor spans share same parent" $ do
        ctx <- newContext
        flip runReaderT ctx . recordSpan (Just Follows) "Foo" $ pure (7 + 8)
        flip runReaderT ctx . recordSpan (Just Follows) "Foo" $ pure (7 + 8)
        [x,y] <- readIORef . spanBuffer $ tracer ctx
        (parentId x == parentId y) @? "Sibling spans must share a parent"
    , testCase "sibling heterogeneous spans share same parent" $ do
        ctx <- newContext
        flip runReaderT ctx . recordSpan (Just Follows) "Foo" $ pure (7 + 8)
        flip runReaderT ctx . recordSpan (Just Child) "Foo" $ pure (7 + 8)
        [x,y] <- readIORef . spanBuffer $ tracer ctx
        (parentId x == parentId y) @? "Sibling spans must share a parent"
    , testCase "sibling spans have distinct ids" $ do
        ctx <- newContext
        flip runReaderT ctx . recordSpan (Just Follows) "Foo" $ pure (7 + 8)
        flip runReaderT ctx . recordSpan (Just Child) "Foo" $ pure (7 + 8)
        [x,y] <- readIORef . spanBuffer $ tracer ctx
        ((sid $ context x) /= (sid $ context y)) @? "Sibling spans must have different ids"
   {- , testCase "nested calls chain" $ do
        ctx <- newContext
        flip runReaderT ctx . recordSpan Nothing "Foo" $
            recordSpan (Just Child) "Bar" $
            recordSpan (Just Child) "Baz"
        -- This test will break if 'recordSpan' changes significantly
        [x, y, z] <- readIORef . spanBuffer $ tracer ctx
        let pid =
        ((sid $ context x) /= (sid $ context y)) @? "Sibling spans must have different ids"
        -}
    ]

newContext :: IO Ctx
newContext = do
    cell <- newIORef []
    currSpan <- newIORef $ SpanId 1234
    pure Ctx {
        tracer = Tracer cell "UNIT TESTING",
        currSpan = currSpan,
        dbg = True,
        currTrace = TraceId 9876
    }

sid :: SpanContext -> SpanId
sid SpanContext {spanId} = spanId

parentId :: Span -> Maybe SpanId
parentId Span {relations=((ChildOf c):_)} = Just $ sid c
parentId Span {relations=((FollowsFrom c):_)} = Just $ sid c
parentId _ =  Nothing

data Ctx = Ctx {
    tracer :: Tracer,
    currSpan :: IORef SpanId,
    dbg :: Bool,
    currTrace :: TraceId
    }

instance (Monad m, MonadBaseControl IO m, MonadIO m, MonadReader Ctx m) => MonadTracer m where
    getTracer = tracer <$> ask
    currentTrace = currTrace <$> ask
    currentSpan = currSpan <$> ask
    isDebug = dbg <$> ask


app/Main.hs
{-# LANGUAGE UndecidableInstances #-}
module Main where

import Servant.Tracing ( getInstructions, WithTracing)
import Tracing.Core (recordSpan,TracingInstructions(..), SpanRelationTag(..), Tracer(..), MonadTracer(..), SpanId(..),
    TraceId(..), debugPrintSpan)
import Tracing.Zipkin (publishZipkin)

import Control.Concurrent (threadDelay, forkIO)
import Control.Monad (forever, mapM_)
import Control.Monad.Trans (MonadIO, liftIO)
import Control.Monad.Trans.Control (MonadBaseControl)
import Control.Monad.Reader (ReaderT(..), ask, MonadReader)
import Data.IORef (IORef, newIORef, atomicModifyIORef')
import Data.Maybe (maybe)
import Data.Proxy (Proxy(..))
import Data.Foldable (toList)
import Data.ByteString.Char8 as BS
import Servant
import Servant.Server
import System.Environment (lookupEnv, getEnv)
import qualified Data.Text as T
import Network.Wai.Handler.Warp (run)
import Network.HTTP.Client (Manager, newManager, defaultManagerSettings, responseStatus, responseBody)


main :: IO ()
main = do
    debug <- maybe False (== "TRUE") <$> lookupEnv "TRACE_DEBUG"
    destinationPath <- getEnv "TRACING_ENDPOINT"
    svcName <- T.pack <$> getEnv "TRACING_SERVICE"
    httpManager <- newManager defaultManagerSettings
    cell <- newIORef []
    let tracer = Tracer cell svcName
    forkIO $ publishLoop destinationPath httpManager tracer
    run 8080 . serve (Proxy :: Proxy ExampleAPI) $ server tracer


publishLoop ::
    String
    -> Manager
    -> Tracer
    -> IO ()
publishLoop destination manager (Tracer {spanBuffer}) = forever $ do
    threadDelay 5000000
    buffer <- atomicModifyIORef' spanBuffer (\b -> ([], b))
    mResp <- publishZipkin destination manager $ toList buffer
    case mResp of
        Nothing -> pure ()
        Just resp -> do
            print $ "Ran Loop " ++ (show $ responseStatus resp) ++ " " ++ (show . fmap debugPrintSpan $ toList buffer)
            print $ "       " ++ (T.unpack $ responseBody resp)



--
-- API Definition
--

type ExampleAPI =
    WithTracing :> MyAPI

type MyAPI =
    Header "Auth" T.Text :>
        (
        "fast" :> Get '[JSON] Int
        :<|>
        "slow" :> Get '[JSON] T.Text
        )


--
-- Server logic
--
server :: Tracer -> Server (WithTracing :> MyAPI)
server tracer inst auth =
    runFast
    :<|>
    runSlow
    where
        loadCtx = do
            instructions <- getInstructions True inst
            currSpan <- liftIO $ newIORef (spanId instructions)
            pure Ctx {
                tracer,
                currSpan,
                instructions
                }
        runFast = do
            ctx <- loadCtx
            runStack ctx $
                recordSpan (const Child <$> inst) "Run Fast" . liftIO $
                    threadDelay 1000000 *> pure 42
        runSlow = do
            ctx <- loadCtx
            runStack ctx $
                recordSpan (const Child <$> inst) "Run Slow" $ do
                    liftIO $ threadDelay 500000
                    let action = liftIO $ threadDelay 1500000 *> pure "Boo"
                    recordSpan (Just Child) "Slow Child" action

runStack :: Ctx -> ReaderT Ctx Handler a -> Handler a
runStack ctx action = runReaderT action ctx

data Ctx = Ctx {
    tracer :: Tracer,
    currSpan :: IORef SpanId,
    instructions :: TracingInstructions
    }

instance (Monad m, MonadBaseControl IO m, MonadIO m, MonadReader Ctx m) => MonadTracer m where
    getTracer = tracer <$> ask
    currentTrace = (traceId . instructions) <$> ask
    currentSpan = currSpan <$> ask
    isDebug = (debug . instructions) <$> ask


src/Tracing/Core.hs
module Tracing.Core (
    Span(..),
    SpanRelation(..),
    SpanRelationTag(..),
    SpanContext(..),
    SpanTag(..),
    OpName(..),
    SpanId(..),
    TraceId(..),
    Tracer(..),
    TracingInstructions(..),
    MonadTracer(..),

    recordSpan,
    debugPrintSpan
    ) where

import Control.Exception.Lifted (bracket)
import Control.Monad.Trans (liftIO, MonadIO)
import Control.Monad.Trans.Control (MonadBaseControl)
import Data.Text (Text)
import qualified Data.Text as T
import Data.Time.Clock (NominalDiffTime, UTCTime, getCurrentTime, diffUTCTime)
import Data.Time.Clock.POSIX (POSIXTime, utcTimeToPOSIXSeconds)
import Data.Int
import Data.Maybe (isJust)
import Data.Monoid ((<>))
import Data.String (IsString)
import System.Random (randomRIO)
import Data.IORef (IORef, atomicModifyIORef',readIORef)
import qualified Data.Map.Strict as M
import Web.HttpApiData (FromHttpApiData)

newtype OpName = OpName Text
    deriving (Eq, Ord, Show, IsString)

newtype SpanId = SpanId Int64
    deriving (Eq, Ord, Show, FromHttpApiData)

newtype TraceId = TraceId Int64
    deriving (Eq, Ord, Show, FromHttpApiData)

-- | Indicates that the current monad can provide a 'Tracer' and related context.
-- It assumes some form of environment. While this exposes some mutable state, all
-- of it is hidden away behind the `recordSpan` api.
class Monad m => MonadTracer m where
    getTracer :: m Tracer
    currentTrace :: m TraceId
    currentSpan :: m (IORef SpanId)
    isDebug :: m Bool

-- | Wraps a computation & writes it to the 'Tracer''s IORef. To start a new top-level span, and therefore
-- a new trace, call this function with *spanType* == 'Nothing'. Otherwise, this will create a child span.
--
-- Doesn't support parallel computations yet
recordSpan :: (MonadIO m, MonadBaseControl IO m, MonadTracer m) =>
    Maybe SpanRelationTag
    -> OpName
    -> m a
    -> m a
recordSpan spanType opName action = do
    Tracer {svcName=serviceName, spanBuffer} <- getTracer
    currentSpanCell <- currentSpan
    activeSpanId <- liftIO $ readIORef currentSpanCell
    traceId <- currentTrace
    debug <- isDebug

    -- generates a thunk that completes once the action provided to 'recordSpan' finishes.
    -- While this is running, there is a new "activeSpanId" that any children will use. Nested calls
    -- generate a stack of spans.
    let startSpan = do
            now <- liftIO getCurrentTime
            newSpanId <- fmap SpanId . liftIO $ randomRIO (0, maxBound)
            let loggedSpanId = resolveSpanId activeSpanId newSpanId
                rel = newSpanRelation traceId activeSpanId
                makeSpan ts =
                    Span {
                        operationName = opName,
                        context = SpanContext traceId loggedSpanId,
                        timestamp = utcTimeToPOSIXSeconds now,
                        relations = rel,
                        tags = M.empty, -- TODO Allow adding these
                        baggage = M.empty, -- TODO Allow adding these
                        duration = diffUTCTime ts now,
                        debug,
                        serviceName
                        }
            liftIO $ atomicModifyIORef' currentSpanCell (const (newSpanId, ()))
            pure $ ActiveSpan makeSpan


        closeSpan (ActiveSpan finishSpan) = do
            now <- liftIO getCurrentTime
            let span = finishSpan now
                sid = spanId (context span :: SpanContext)
            liftIO $ atomicModifyIORef' spanBuffer (\xs -> (span:xs, ()))
            liftIO $ atomicModifyIORef' currentSpanCell (const (activeSpanId, ()))

    bracket startSpan
            closeSpan
            (const action)

    where
        -- When this is a top level span, there should be no SpanRelationTag. These two functions work
        -- together to ensure the spans nest properly
        resolveSpanId activeSpanId newSpanId =
            if isJust spanType
            then newSpanId
            else activeSpanId
        newSpanRelation traceId activeSpanId =
            case spanType of
                Just Child -> [ChildOf $ SpanContext traceId activeSpanId]
                Just Follows -> [FollowsFrom $ SpanContext traceId activeSpanId]
                Nothing -> []

-- | Instructions that are specific to a single trace
data TracingInstructions =
    TracingInstructions {
        traceId :: !TraceId,
        spanId :: !SpanId,
        parentSpanId :: !SpanId,
        debug :: !Bool,
        sample :: !Bool
    } deriving (Eq, Show)

newtype ActiveSpan =
    ActiveSpan {finishSpan :: UTCTime -> Span}

-- | Global context required for tracing
data Tracer =
    Tracer {
        spanBuffer :: IORef [Span],
        svcName :: T.Text
    }

data SpanContext =
    SpanContext {
        traceId :: !TraceId,
        spanId :: !SpanId
    } deriving (Eq, Show)

data SpanRelation =
    ChildOf !SpanContext | FollowsFrom !SpanContext
    deriving (Eq, Show)

data SpanRelationTag = Child | Follows

data Span = Span {
    operationName :: !OpName,
    context :: !SpanContext,
    timestamp :: !POSIXTime,
    duration :: !NominalDiffTime,
    relations :: ![SpanRelation],
    tags :: !(M.Map Text SpanTag),
    baggage:: !(M.Map Text Text),
    debug :: !Bool,
    serviceName :: !Text
    } deriving Show

debugPrintSpan ::
    Span
    -> Text
debugPrintSpan span =
    "Span: " <>
    "id ["<>(unSpan $ spanId (context span :: SpanContext))<>"] "<>
    "op ["<>(unOp $ operationName span)<>"] "<>
    "duration ["<>(T.pack . show $ duration span)<> "] "<>
    "relations "<>(T.pack . show $ relations span)
    where
        unOp (OpName o) = o
        unSpan (SpanId s) = T.pack $ show s

data SpanTag
    = TagString !Text
    | TagBool !Bool
    | TagInt !Int64
    deriving (Eq, Show)


src/Tracing/Jaeger.hs
{-# LANGUAGE ScopedTypeVariables #-}

module Tracing.Jaeger (
    publishJaeger,
    ZipkinSpan(..)
    ) where

import Tracing.Core (Span(..), SpanId(..), OpName(..), TraceId(..), SpanContext(..),
    SpanRelation(..))

import Control.Monad.Trans (liftIO, MonadIO)
import Control.Monad (void)
import Data.Monoid ((<>), mempty)
import Data.Aeson
import Data.Maybe (fromMaybe)
import Data.Int (Int64)
import qualified Data.Text as T
import qualified Data.Text.Encoding as T
import qualified Data.ByteString.Char8 as BS
import qualified Data.ByteString.Lazy as LBS
import qualified Data.ByteString.Lex.Integral as BS
import qualified Data.HashMap.Strict as HM
import Network.HTTP.Client


-- | Forces a dependency on `Req`
publishJaeger :: MonadIO m =>
    String
    -> Manager
    -> [Span]
    -> m (Maybe (Response T.Text))
publishJaeger _ _ [] = pure Nothing
publishJaeger destination manager spans =
    liftIO . fmap (Just . fmap decode) $ httpLbs jaegerReq manager
    where
        decode = T.decodeUtf8 . LBS.toStrict
        req = parseRequest_ destination
        body = RequestBodyLBS . encode $ ZipkinSpan <$> spans
        jaegerReq = req { method = "POST", requestBody = body, requestHeaders = [("content-type", "application/json")]}

newtype ZipkinSpan = ZipkinSpan Span
instance ToJSON ZipkinSpan where
    toJSON (ZipkinSpan span) = object $ [
        "traceId" .= unTrace (traceId $ context span),
        "id" .=  unSpan (spanId $ context span),
        "name" .=  unOp (operationName span),
        "timestamp" .= (floor . toMicros $ timestamp span :: Int64),
        "kind" .= ("CLIENT"::T.Text),
        "duration" .= (toMicros $ duration span),
        "debug" .= (debug span),
        "localEndpoint" .= (object ["serviceName" .= (serviceName span)])
        ] <>
        parentId (relations span)
        where
            toMicros = (*) 1000000
            unOp (OpName n) = n
            zipkinFormatId = padLeft 16 . T.pack . BS.unpack . fromMaybe "-1"
            unTrace (TraceId t) = zipkinFormatId $ BS.packHexadecimal t
            unSpan (SpanId s) = zipkinFormatId $ BS.packHexadecimal s
            parentId :: [SpanRelation] -> [(T.Text, Value)]
            parentId (ChildOf ctx:_) = ["parentId" .= (unSpan $ spanId ctx)]
            parentId (FollowsFrom ctx:_) = ["parentId" .= (unSpan $ spanId ctx)]
            parentId _ = []
            padLeft 0 txt = txt
            padLeft n txt
                | T.length txt < n = padLeft n ("0"<>txt)
                | otherwise = txt


src/Servant/Tracing.hs
{-# LANGUAGE RankNTypes #-}
module Servant.Tracing (
    ServantTracingT,
    WithTracing,
    TracingInstructions(..),
    instructionsToHeader,
    getInstructions
    ) where

import Tracing.Core (Tracer, TraceId(..), SpanId(..), MonadTracer, TracingInstructions(..))

import Control.Monad.Trans (liftIO, MonadIO)
import qualified Data.Text as T
import qualified Data.ByteString.Char8 as BS
import qualified Data.ByteString.Lex.Integral as BS
import Data.Text.Read(hexadecimal)
import Data.Bits (testBit, (.|.))
import Data.Maybe (catMaybes, fromMaybe)
import Data.Monoid ((<>))
import Servant
import Servant.Server
import System.Random (randomRIO)
import Web.HttpApiData (FromHttpApiData(..))

-- | Constrain the 'ServerT''s base monad such that it provides an instance of 'MonadTracer'
type ServantTracingT api m = (MonadIO m, MonadTracer m) => ServerT api m
type WithTracing = Header "uber-trace-id" TracingInstructions


-- | Jaeger format: http://jaeger.readthedocs.io/en/latest/client_libraries/#propagation-format
instructionsToHeader :: TracingInstructions -> T.Text
instructionsToHeader TracingInstructions {traceId=(TraceId tid), spanId=SpanId sid, parentSpanId=SpanId pid, sample, debug} =
    toField tid<>":"<> toField  sid <> ":"<> toField pid <> ":" <> (T.pack $ show setFlags)
    where
        toField = T.pack . BS.unpack . fromMaybe "0" . BS.packHexadecimal
        setFlags :: Int
        setFlags = (if debug then 2 else 0) .|. (if sample then 1 else 0) .|. 0


instance FromHttpApiData TracingInstructions where
    parseUrlPiece ::
        T.Text
        -> Either T.Text TracingInstructions
    parseUrlPiece raw =
        case T.splitOn ":" raw of
            [rawTraceId, rawSpanId, rawParentId, flags] -> let
                res = do
                    traceId <- TraceId . fromIntegral . fst <$> hexadecimal rawTraceId
                    spanId <- SpanId . fromIntegral . fst <$> hexadecimal rawSpanId
                    parentId <- SpanId . fromIntegral . fst <$> if T.null rawParentId then pure (0 :: Integer, "") else hexadecimal rawParentId
                    flagField <- fromIntegral . fst <$> hexadecimal flags
                    let [sample, debug]= [sampleFlag, debugFlag] <*> [flagField]
                    pure TracingInstructions {
                       traceId = traceId,
                       spanId = spanId,
                       parentSpanId = parentId,
                       sample = sample,
                       debug = debug
                    }
                in case res of
                    Left err -> Left $ T.pack err
                    Right val -> Right  val
            _ -> Left $ raw <> " is not a valid uber-trace-id header"
        where
            sampleFlag :: Int -> Bool
            sampleFlag = (`testBit` 0)
            debugFlag :: Int -> Bool
            debugFlag = (`testBit` 1)


-- TODO write a monad that wraps servant & determines if it should sample or not. Takes a sampling determinant. Only evaluates if the header is not present
getInstructions :: MonadIO m =>
    Bool
    -> Maybe TracingInstructions
    -> m TracingInstructions
getInstructions debug Nothing = do
    newTraceId <- liftIO $ randomRIO (0, maxBound)
    newSpanId <- liftIO $ randomRIO (0, maxBound)
    sample <- liftIO $ randomRIO (0, 1000)
    pure TracingInstructions {
        traceId = TraceId newTraceId,
        spanId = SpanId newSpanId,
        parentSpanId = SpanId 0,
        debug,
        sample = sample == (1::Int)
        }
getInstructions _ (Just inst) = pure inst


src/Tracing/Zipkin.hs
{-# LANGUAGE ScopedTypeVariables #-}

module Tracing.Zipkin (
    publishZipkin,
    ZipkinSpan(..)
    ) where

import Tracing.Core (Span(..), SpanId(..), OpName(..), TraceId(..), SpanContext(..),
    SpanRelation(..))

import Control.Monad.Trans (liftIO, MonadIO)
import Control.Monad (void)
import Data.Monoid ((<>), mempty)
import Data.Aeson
import Data.Maybe (fromMaybe)
import Data.Int (Int64)
import qualified Data.Text as T
import qualified Data.Text.Encoding as T
import qualified Data.ByteString.Char8 as BS
import qualified Data.ByteString.Lazy as LBS
import qualified Data.ByteString.Lex.Integral as BS
import qualified Data.HashMap.Strict as HM
import Network.HTTP.Client


-- | Publish 'Span' in the Zipkin format (TODO add link to zipkin spec here). No call is made
-- on an empty
publishZipkin :: MonadIO m =>
    String -- The address of the backend server
    -> Manager
    -> [Span] -- ^ The traced spans to send to a Zipkin backend
    -> m (Maybe (Response T.Text))
publishZipkin _ _ [] = pure Nothing
publishZipkin destination manager spans =
    liftIO . fmap (Just . fmap decode) $ httpLbs zipkinReq manager
    where
        decode = T.decodeUtf8 . LBS.toStrict
        req = parseRequest_ destination
        body = RequestBodyLBS . encode $ ZipkinSpan <$> spans
        zipkinReq = req { method = "POST", requestBody = body, requestHeaders = [("content-type", "application/json")]}

newtype ZipkinSpan = ZipkinSpan Span
instance ToJSON ZipkinSpan where
    toJSON (ZipkinSpan span) = object $ [
        "traceId" .= unTrace (traceId $ context span),
        "id" .=  unSpan (spanId $ context span),
        "name" .=  unOp (operationName span),
        "timestamp" .= (floor . toMicros $ timestamp span :: Int64),
        "kind" .= ("CLIENT"::T.Text),
        "duration" .= (toMicros $ duration span),
        "debug" .= (debug span),
        "localEndpoint" .= (object ["serviceName" .= (serviceName span)])
        ] <>
        parentId (relations span)
        where
            toMicros = (*) 1000000
            unOp (OpName n) = n
            zipkinFormatId = padLeft 16 . T.pack . BS.unpack . fromMaybe "-1"
            unTrace (TraceId t) = zipkinFormatId $ BS.packHexadecimal t
            unSpan (SpanId s) = zipkinFormatId $ BS.packHexadecimal s
            parentId :: [SpanRelation] -> [(T.Text, Value)]
            parentId (ChildOf ctx:_) = ["parentId" .= (unSpan $ spanId ctx)]
            parentId (FollowsFrom ctx:_) = ["parentId" .= (unSpan $ spanId ctx)]
            parentId _ = []
            padLeft 0 txt = txt
            padLeft n txt
                | T.length txt < n = padLeft n ("0"<>txt)
                | otherwise = txt
